src/lib/container/mqueue.h

```diff
@@ -0,0 +1,38 @@
+/* Copyright (c) 2003-2004, Roger Dingledine
+ * Copyright (c) 2004-2006, Roger Dingledine, Nick Mathewson.
+ * Copyright (c) 2007-2018, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+#ifndef TOR_MQUEUE_H
+#define TOR_MQUEUE_H
+
+/**
+ * \file mqueue.h
+ *
+ * \brief Header for mqueue.c
+ **/
+
+#include <stddef.h>
+
+struct smartlist_t;
+/**
+ * A message queue, backed by a pair of smartlists.
+ */
```

Comment:
I'm not sure I understand why there needs to be this additional complexity of using a pair of smartlists. Is this to work around the limitations of smartlists?

Reply:
Hm. I think maybe it's worth it to go with a simpler implementation here. I have a slight preference for taking a O(n) performance hit to delete from the front of the array, under the assumption that n will tend to be small. At a minimum there's already a penalty of O(n) to do the reversal whenever the "flushing" smartlist is empty and we need to refill it. (Also `memmove()` is likely better optimized than our code that does smartlist reversal. Then again, with the single smartlist approach, we would potentially call `memmove()` for each pop compared to reversing the list fewer times in the dual smartlist approach.)

If it turns out that we do need O(1) amortized pop-from-beginning, we can implement a circular queue. Are there other places in tor that could take advantage of a ring buffer or circular queue?

Do we already have expected use cases for pubsub where we expect large queues? I'm not sure.

I'm undecided whether the dual smartlist approach is more or less risky than a circular queue in terms of complexity.

Reply:
I've done two slightly different ringbuf implementations in "messaging_ringbuf" and "messaging_ringbuf_v2".  I don't vouch for their correctness: much more testing and analysis would be needed.  Nonetheless, they might help come up with a picture of how the complexity differs between the two approaches.

Reply:
Summarizing some IRC conversation: using the BSD queue macros (maybe SIMPLEQ?) seems like a logical thing here: always O(1) enqueue and dequeue. The downsides are that they involve intrusive structures. `msg_t` is visible to at least the receiving functions in subscribers, so the intrusive link fields would be visible to them as well. We don't gain much locality of reference from using an array-based implementation, given that it's arrays of pointers to individually heap-allocated objects.

Wrapping a `msg_t` in another struct might be one way to hide the links from the consuming code, assuming that only approved functions get used to allocate or free these objects. This is similar to an approach I've seen elsewhere in tor.

I think it's less complicated to allow the link structures to be intrusive, so I think we should do that unless there's a good reason. (Do we need to be defensive against other tor developers inappropriately using the link fields in a publisher or subscriber? Data hiding comes with a complexity cost, and C isn't a good language for hiding that complexity.)